File: app/src/main/java/com/doapps/habits/helper/DayFilter.java
package com.doapps.habits.helper;

import android.util.Log;

import com.doapps.habits.models.Habit;

import java.util.Calendar;
import java.util.Iterator;

class DayFilter {

    static void filterListForToday(Iterable<Habit> todayHabits) {
        int dayOfWeek = Calendar.getInstance().get(Calendar.DAY_OF_WEEK);
        Iterator<Habit> habitIterator = todayHabits.iterator();
        while (habitIterator.hasNext()) {
            int[] freq = habitIterator.next().getFrequencyArray();
            if (freq.length == 0) {
                Log.w("filterListByDay()", "frequency not set");
            } else {
                if (freq.length > 2) { // not once type
                    boolean today = false;
                    for (int i = 1; i < freq.length; i++) {
                        if (freq[i] == dayOfWeek) {
                            today = true;
                            break;
                        }
                    }
                    if (!today) {
                        // if not found days markers
                        habitIterator.remove();
                    }
                }
            }
        }
    }

    static void filterListByDay(Iterable<Habit> dayHabits, int dayOfWeek) {
        Iterator<Habit> habitIterator = dayHabits.iterator();
        while (habitIterator.hasNext()) {
            int[] freq = habitIterator.next().getFrequencyArray();
            if (freq.length == 0) {
                Log.w("filterListByDay()", "frequency not set");
            } else if (!(freq.length == 2 && freq[0] == freq[1])) {
                // if it's not every day
                if (freq.length > 2) {
                    boolean today = false;
                    for (int j = 1; j < freq.length; j++) {
                        if (freq[j] == dayOfWeek) {
                            today = true;
                            break;
                        }
                    }
                    if (!today) {
                        // if not found days markers
                        habitIterator.remove();
                    }
                } else {
                    // once type
                    habitIterator.remove();
                }
            }
        }
    }
}

File: app/src/main/java/com/doapps/habits/helper/DayFilter.kt
package com.doapps.habits.helper

import android.util.Log
import com.doapps.habits.BuildConfig
import com.doapps.habits.models.Habit
import java.util.*

object DayFilter {

  @JvmStatic
  fun filterListForToday(todayHabits: MutableCollection<Habit>) {
    val dayOfWeek = Calendar.getInstance().get(Calendar.DAY_OF_WEEK)
    val habitIterator = todayHabits.iterator()
    while (habitIterator.hasNext()) {
      val freq = habitIterator.next().frequencyArray

      if (!freq.isEmpty() && freq.size > 2 && notToday(freq, dayOfWeek)) habitIterator.remove()
      else if (BuildConfig.DEBUG) Log.w("filterListByDay()", "frequency not set")
    }
  }

  @JvmStatic
  fun filterListByDay(dayHabits: MutableCollection<Habit>, dayOfWeek: Int) {
    val habitIterator: MutableIterator<Habit> = dayHabits.iterator()
    while (habitIterator.hasNext()) {
      val freq = habitIterator.next().frequencyArray

      if (!freq.isEmpty() && !isEveryday(freq) && (once(freq) || notToday(freq, dayOfWeek))) habitIterator.remove()
      else if (BuildConfig.DEBUG) Log.w("filterListByDay()", "frequency not set")
    }
  }

  private fun once(freq: IntArray) = freq.size <= 2
  private fun notToday(freq: IntArray, dayOfWeek: Int) = (1 until freq.size).none { freq[it] == dayOfWeek }
  private fun isEveryday(freq: IntArray) = freq.size == 2 && freq[0] == freq[1]
}
